### PR TITLE
Add fnc_insert

### PR DESCRIPTION
**Pull request by 654wak654**
**When merged this pull request will:**
- Title

### CODE REVIEW

**Comment by BaerMitUmlaut** [comment]
You need to add it to CfgFunctions too.

**Comment by 654wak654** [comment]
:picard:^2

File: addons/arrays/fnc_insert.sqf [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+#include "script_component.hpp"
+SCRIPT(insert);
+
+params ["_array", "_index", "_element"];
+
+private _size = count _array;
+if (_index >= _size || {_index < 0}) exitWith {[]};
+
+_right = _array select [_index, _size];
+_array deleteRange [_index, _size];
```
Comment (commy2) [inline addons/arrays/fnc_insert.sqf]: `_array resize [_index - 1];`

File: addons/arrays/fnc_insert.sqf [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+---------------------------------------------------------------------------- */
+#include "script_component.hpp"
+SCRIPT(insert);
+
+params ["_array", "_index", "_element"];
+
+private _size = count _array;
+if (_index >= _size || {_index < 0}) exitWith {[]};
+
+_right = _array select [_index, _size];
```
Comment (commy2) [inline addons/arrays/fnc_insert.sqf]: missing `private`

File: addons/arrays/fnc_insert.sqf [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+Author:
+    654wak654
+---------------------------------------------------------------------------- */
+#include "script_component.hpp"
+SCRIPT(insert);
+
+params ["_array", "_index", "_element"];
+
+private _size = count _array;
+if (_index >= _size || {_index < 0}) exitWith {[]};
```
Comment (commy2) [inline addons/arrays/fnc_insert.sqf]: `[[], 1, "test"] call CBA_fnc_insert`
should report
`[nil, "test"]`
imo

Comment (654wak654) [inline addons/arrays/fnc_insert.sqf]: So just `exitWith {}` or `exitWith {nil}`?

**Comment by commy2** [comment]
If we'd replace `_element` with an array to insert, all we would have to change is one `pushBack` to `append`.
If you wanted to insert a single element, instead of
`[_arr, 2, "c"] call CBA_fnc_insert;`
you would write:
`[_arr, 2, ["c"]] call CBA_fnc_insert;`
, but you would still have the option to insert multiple elements at once.

**Comment by commy2** [comment]
There exists `BIS_fnc_arrayInsert`

```sqf
/************************************************************
	Array Insert
	Author: Andrew Barron, optimised by Killzone_Kid

Parameters: [base array, insert array, index]
Returs: [array]

Inserts the elements of one array into another, at a specified
index.

Neither arrays are touched by reference, a new array is returned.

Example1: [[0,1,2,3,4], ["a","b","c"], 1] call BIS_fnc_arrayInsert
Returns: [0,"a","b","c",1,2,3,4]
Example2: [[1,2],[3,4],5] call BIS_fnc_arrayInsert
Returns: [1,2,<null>,<null>,<null>,3,4]
************************************************************/

/// --- validate general input
#include "..\paramsCheck.inc"
#define arr [[],[],0]
paramsCheck(_this,isEqualTypeParams,arr)

params ["_arr1", "_arr2", "_index"];

if (_index < 0) exitWith {_arr1};

private _size1 = count _arr1;
if (_index >= _size1) then {
	_arr1 = +_arr1;
	_arr1 resize _index;
};

(_arr1 select [0, _index]) + _arr2 + (_arr1 select [_index, _size1])
```

Differences:
- has input validation
- creates new array instead of modifying it
- slightly different order of parameters

We now have to think about:
- how does our variant perform compared to the (KK rewritten) BI function?
- do we need this CBA function (usefull feature vs. bloat?)

**Comment by commy2** [comment]
@654wak654 Sorry for the delay. Can you add a unit test file for this? Would've to be added here:
https://github.com/CBATeam/CBA_A3/blob/master/addons/arrays/test.sqf

**Pull request merged**